src/main/java/org/tarantool/TarantoolClientImpl.java

```diff
@@ -147,6 +152,9 @@ protected void reconnect(Throwable lastError) {
         int retryNumber = 0;
         while (!Thread.currentThread().isInterrupted()) {
             try {
+                if (lastError != NOT_INIT_EXCEPTION) {
+                    LOGGER.warn("Attempt to (re-)connect to Tarantool instance", lastError);
```

Comment:
it depends on the conventions we follow.
For instance, I prefer the following list:
1. The `ERROR` level should only be used when the application really is in trouble. Users are being affected without having a way to work around the issue.
2. The `WARN` level should be used when something bad happened, but the application still has the chance to heal itself or the issue can wait a day or two to be fixed.
3. The INFO level should be used to document state changes in the application or some entity within the application.
4. In a nutshell, we want to log any information that helps us identify what went wrong on `DEBUG` level.
5. When `TRACE` is used, we want to log all information that helps us to trace the processing of an incoming request through our application.

Of course, we can consider other points that work for us.

Reply:
I enabled the warning manually and it was expected to get it. The questions above are about how the warning should look: should it contain a reason of a reconnection, should it contain a remote address?

Reply:
Looks reasonable, but the remote address is socket provider internals (in case of RoundRobin it's next available address from the pool). I think we can extend SocketProvider interface to get such info.

Reply:
Is just SocketChannel#getRemoteAddress() sufficient to obtain an address?

Reply:
No, we have SocketChannel object after successful connection after #get(int, Throwable) call. But we have to log before the attempt.

Reply:
Now I see, thanks. Okay.

Reply:
> Why it does not show an error?

Seems you used my default pattern to log. I added the missed error object and now It looks like `[%1$tF %1$tT] [%4$-7s] %3$s %5$s%6$s %n`.

> It worth to write an address of an instance: host:port or username:***@host:port.

Done.

Reply:
Looks strange, LOGGER.warn() works if you choose WARNING and value below (see [api](https://docs.oracle.com/javase/7/docs/api/java/util/logging/Level.html)). So, there are two values SEVERE and OFF that disable warning messages. In your case (commented lines) it means that you are gonna use the default level which is INFO (it also includes WARN).

Reply:
Found the cause:

```diff
diff --git a/src/main/java/org/tarantool/logging/BaseLogger.java b/src/main/java/org/tarantool/logging/BaseLogger.java
index 81967e9..8a108f8 100644
--- a/src/main/java/org/tarantool/logging/BaseLogger.java
+++ b/src/main/java/org/tarantool/logging/BaseLogger.java
@@ -112,7 +112,7 @@ abstract class BaseLogger implements Logger {
 
     @Override
     public void warn(Supplier<String> message, Throwable throwable) {
-        log(Level.TRACE, throwable, message);
+        log(Level.WARN, throwable, message);
     }
 
     @Override
```

However after the change (and enabling logging as shown above) I observed the following failures:

```
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   ClientReconnectIT.testConcurrentCloseAndReconnect:231 expected: <true> but was: <false>
[ERROR] Errors: 
[ERROR]   ClientReconnectClusterIT.testDelayFunctionResultFetch:430->expectConnected:468 » Communication
[ERROR]   ClientReconnectIT.testReconnect:106 » IllegalState State is CLOSED.
[ERROR]   ClientReconnectIT.testReconnectWhileOperationsAreInProgress:201 » IllegalState
[INFO] 
[ERROR] Tests run: 345, Failures: 1, Errors: 3, Skipped: 7
```

Are they reproduced for you?

Reply:
fixed